src/pow/work.rs

```diff
@@ -42,19 +43,44 @@ impl Work {
 
     /// Block and generate forever until we find a solution.
     pub fn generate(subject: &Subject, threshold: &Difficulty) -> anyhow::Result<Work> {
+        let mut work_and_subject = [0u8; 40];
+
+        // We can place the subject in the second part of the slice which will not change.
+        let subject_slice = &mut work_and_subject[Self::LEN..];
+
+        // This will panic if the source slice is too big. The assertion here is a safety measure
+        // in debug builds.
+        debug_assert_eq!(subject.as_bytes().len(), subject_slice.len());
+        subject.as_bytes().copy_to_slice(subject_slice);
+
+        let mut difficulty: Difficulty = Difficulty::new(0);
+
+        // Fill the first 8 bytes with the random work.
+        let work_slice = &mut work_and_subject[0..Self::LEN];
+        rand::thread_rng().fill_bytes(work_slice);
+
         loop {
-            let work = Work::attempt(&subject, &threshold)?;
-            if work.is_none() {
-                continue;
+            // Pick a random byte position and increment.
+            // I'm guessing this is slightly faster than using fill_bytes for a new set of numbers.
+            // TODO: Bench this guess.
+            let idx = (rand::random::<u8>() % (Self::LEN as u8)) as usize;
```

Comment:
I would even go as far to suggest just deterministically increment by 1. I'm going to assume the blake2b generates a uniform distribution by doing that, so generating a random number is redundant. From wikipedia:
> Changing a single bit causes each bit in the output to change with 50% probability

Reply:
I did want to increment by one, but I couldn't think of a ("safe") way to overflow each digit with high performance. i.e. I would have have to convert to u64 with an overflow check then back to a byte slice.